tests/semantic_test.py

```diff
@@ -0,0 +1,561 @@
+from pathlib import Path
+from itertools import product
+from typing import Dict, Tuple
+
+import pytest
+
+from .compile_asserts import assert_teal_as_expected
+from .semantic_asserts import (
+    algod_with_assertion,
+    get_blackbox_scenario_components,
+    e2e_pyteal,
+    mode_has_assertion,
+)
+
+from algosdk.testing.dryrun import Helper as DryRunHelper
+from algosdk.testing.teal_blackbox import (
+    DryRunAssertionType as DRA,
+    SequenceAssertion,
+    csv_from_dryruns,
+    dryrun_assert,
+    lightly_encode_args,
+    lightly_encode_output,
+    scratch_encode,
+)
+
+from pyteal import *
+
+# TODO: get tests working on Circle and set this to True
+SEMANTIC_TESTING = True
+# TODO: remove thise skips after the following issue has been fixed https://github.com/algorand/pyteal/issues/199
+STABLE_SLOT_GENERATION = False
+SKIP_SCRATCH_ASSERTIONS = not STABLE_SLOT_GENERATION
+
+
+####### Subroutine Definitions Being Tested ########
+
+
+@Subroutine(TealType.uint64, input_types=[])
+def exp():
+    return Int(2) ** Int(10)
+
+
+@Subroutine(TealType.none, input_types=[TealType.uint64])
+def square_byref(x: ScratchVar):
+    return x.store(x.load() * x.load())
+
+
+@Subroutine(TealType.uint64, input_types=[TealType.uint64])
+def square(x):
+    return x ** Int(2)
+
+
+@Subroutine(TealType.none, input_types=[TealType.anytype, TealType.anytype])
+def swap(x: ScratchVar, y: ScratchVar):
+    z = ScratchVar(TealType.anytype)
+    return Seq(
+        z.store(x.load()),
+        x.store(y.load()),
+        y.store(z.load()),
+    )
+
+
+@Subroutine(TealType.bytes, input_types=[TealType.bytes, TealType.uint64])
+def string_mult(s: ScratchVar, n):
+    i = ScratchVar(TealType.uint64)
+    tmp = ScratchVar(TealType.bytes)
+    start = Seq(i.store(Int(1)), tmp.store(s.load()), s.store(Bytes("")))
+    step = i.store(i.load() + Int(1))
+    return Seq(
+        For(start, i.load() <= n, step).Do(s.store(Concat(s.load(), tmp.load()))),
+        s.load(),
+    )
+
+
+@Subroutine(TealType.uint64, input_types=[TealType.uint64])
+def oldfac(n):
+    return If(n < Int(2)).Then(Int(1)).Else(n * oldfac(n - Int(1)))
+
+
+@Subroutine(TealType.uint64, input_types=[TealType.uint64])
+def slow_fibonacci(n):
+    return (
+        If(n <= Int(1))
+        .Then(n)
+        .Else(slow_fibonacci(n - Int(2)) + slow_fibonacci(n - Int(1)))
+    )
+
+
+def fac_with_overflow(n):
+    if n < 2:
+        return 1
+    if n > 20:
+        return 2432902008176640000
+    return n * fac_with_overflow(n - 1)
+
+
+def fib(n):
+    a, b = 0, 1
+    for _ in range(n):
+        a, b = b, a + b
+    return a
+
+
+def fib_cost(args):
+    cost = 17
+    for n in range(1, args[0] + 1):
+        cost += 31 * fib(n - 1)
+    return cost
+
+
+APP_SCENARIOS = {
```

Comment:
cf. [PySDK's PR #298](https://github.com/algorand/py-algorand-sdk/blob/45f4001a7145c3526ac11d88b73b112ab57151ca/x/blackbox/README.md) for an explanation of what this all means